File: backend/dataset/maxPathSum/maxPathSum_18.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxPathSum(self, root:TreeNode)->int:
        """
        :type root: TreeNode
        :rtype: int
        """
        self.res = float('-inf')
        self.helper(root)
        return self.res 
        
    def helper(self, root):
        if not root:
            return 0
        left, right = self.helper(root.left), self.helper(root.right)
        self.res = max(self.res, root.val + left + right)
        return max(root.val + max(left, right), 0)

File: backend/CDataset/maxPathSum/maxPathSum_18.cpp
class Solution {
public:
    #include<vector>
    class TreeNode{
        public:
        TreeNode* left;
        TreeNode* right;
        int val;
    };
	int maxPathSum(TreeNode* root) {
	int res = INT_MIN;
	dfs(root, res);
	return res;
}

int dfs(TreeNode *node, int &res) {
	if(!node) return 0;
	int l = dfs(node->left, res), r = dfs(node->right, res);
	res = max(res, node->val + l + r);
	return max(0, node->val + max(l, r));
}
};